pkg/boot/universalpayload/universalpayload.go

```diff
@@ -0,0 +1,449 @@
+	// This hardcode is introduced by golang structure algnment vs. write bytes.
+	var pad [4]byte
+	if err := binary.Write(buf, binary.LittleEndian, pad); err != nil {
+		return fmt.Errorf("failed to append smbios table pad to buffer")
+	}
+
+	if length != (uint64)(buf.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending smbios table")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func appendEfiCPUHob(buf *bytes.Buffer, hobLen *uint64) error {
+	cpuHob, err := hobCreateEfiHobCPU()
+	if err != nil {
+		return err
+	}
+
+	length := uint64(unsafe.Sizeof(EfiHobCPU{}))
+	prev := buf.Len()
+	if err := binary.Write(buf, binary.LittleEndian, cpuHob); err != nil {
+		return fmt.Errorf("failed to append cpu hob to buffer")
+	}
+
+	if length != (uint64)(buf.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending cpu hob")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func constructHobList(dst *bytes.Buffer, src *bytes.Buffer, hobLen *uint64) error {
+	handoffHob := hobCreateEfiHobHandoffInfoTable(*hobLen)
+	if err := binary.Write(dst, binary.LittleEndian, handoffHob); err != nil {
+		return fmt.Errorf("failed to append handoff hob to buffer")
+	}
+
+	if err := binary.Write(dst, binary.LittleEndian, src.Bytes()); err != nil {
+		return fmt.Errorf("failed to append hos list to buffer")
+	}
+
+	hobEndHeader := hobCreateEndHob()
+	prev := dst.Len()
+	length := uint64(unsafe.Sizeof(EfiHobGenericHeader{}))
+
+	if err := binary.Write(dst, binary.LittleEndian, hobEndHeader); err != nil {
+		return fmt.Errorf("failed to append end hob")
+	}
+
+	if length != (uint64)(dst.Len()-prev) {
+		return fmt.Errorf("length mismatch when appending end hob")
+	}
+
+	*hobLen += length
+
+	return nil
+}
+
+func loadKexecMemWithHobs(fdtLoad *FdtLoad, fdtData []byte) (kexec.Memory, error) {
+	//Step 1, Prepare memory
+	mem := kexec.Memory{}
+
+	//Step 2, Insert tianocore raw binary
+	mem.Segments.Insert(kexec.NewSegment(fdtData, kexec.Range{Start: uintptr(fdtLoad.Load), Size: uint(len(fdtData))}))
+
+	// Step 3, Prepare HobList
+	// TODO: remove hardcode HoB Address here
+	hobAddr := fdtLoad.Load - 0x100000
+	hobBuf := &bytes.Buffer{}
+	hobListBuf := &bytes.Buffer{}
+	var hobLen uint64
+
+	if err := appendMemMapHob(hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	if err := appendSerialPortHob(hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	if err := appendUniversalPayloadBase(hobBuf, &hobLen, fdtLoad.Load); err != nil {
+		return mem, err
+	}
+
+	if err := appendAcpiTableHob(hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	if err := appendSmbiosTableHob(hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	if err := appendEfiCPUHob(hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	if err := constructHobList(hobListBuf, hobBuf, &hobLen); err != nil {
+		return mem, err
+	}
+
+	mem.Segments.Insert(kexec.NewSegment(hobListBuf.Bytes(), kexec.Range{Start: uintptr(hobAddr), Size: uint(hobLen)}))
```

Comment:
ditto, how do we know this is a usable piece of memory?

Reply:
Same as above reply.